falguard.py

```diff
@@ -69,13 +69,28 @@ def __init__(self, spec_path):
     def __call__(self, request, response, resource, params):
         self.process_resource(request, response, resource, params)
 
+    def _build_http_not_allowed_error(self, uri):
+        # Error 405 requires we provide the list of allowed methods
+        # for this URI. If None is found, then we produce a 404 error.
+        allowed_methods = [method for method in falcon.HTTP_METHODS
+                           if self._spec.get_op_for_request(method, uri)]
+
+        return (falcon.HTTPMethodNotAllowed(allowed_methods) if allowed_methods
+                else falcon.HTTPNotFound)
```

Comment:
This whole private method looks a bit unnecessary. My suggestion for refactoring right in the process_resource method:

```python
        uri = request.uri_template
        if not uri:
            raise falcon.HTTPNotFound

        method = request.method.upper()
        get_op = functools.partial(
            self._spec.get_op_for_request,
            path_pattern=uri,
        )

        operation = get_op(method)
        if not operation:
            allowed = [
                other_method for other_method in falcon.HTTP_METHODS
                if other_method != method and get_op(other_method)
            ]
            if allowed:
                raise falcon.HTTPMethodNotAllowed(allowed)
            raise falcon.HTTPNotFound
```